walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,151 @@
+        // See if they are attending the event
+        if (e.getAttendees().contains(guest)){
+          // Get the time of the event so we can mark as busy
+          guestBusy.add(e.getWhen());
+        }
+      }
+      busyTimes.add(guestBusy);
+    }
+    
+    // Use our busy times list to find all times when our guests are NOT busy
+    ArrayList<ArrayList<TimeRange>> allTimes = new ArrayList<>();
+    for (ArrayList<TimeRange> guestSchedule : busyTimes){
+      allTimes.add(findAvailability(guestSchedule, duration));
+    }
+    return findIntersectionBetweenLists(allTimes, duration);
+  }
+
+  /* 
+   Given a guest's schedule, finds all open times during schedule
+   where a meeting could take place.
+   @param guestSchedule Takes a list of when guests are busy
+   @param duration The length of the desired meeting
+   @return ArrayList<TimeRange> Returns a list containing the TimeRanges where the meeting could occur
+  */
+  public ArrayList<TimeRange> findAvailability(ArrayList<TimeRange> guestSchedule, int duration){
+    ArrayList<TimeRange> available = new ArrayList<>();
+    // If their schedule is empty, they are free the entire day
+    if (guestSchedule.size() == 0){
+      available.add(TimeRange.WHOLE_DAY);
+    } else {
+      Collections.sort(guestSchedule, TimeRange.ORDER_BY_START);
+      int st = 0;
+      TimeRange t = null;
+      // Find available time ranges between each event in their schedule
+      for (int i = 0; i < guestSchedule.size(); i++){
+        t = guestSchedule.get(i);
+        if (st < t.start() && t.start() - st >= duration){
+          TimeRange possible = TimeRange.fromStartEnd(st, t.start(), false);
+          boolean modified = false;
+          // Accounts for double-booked schedules. 
+          for (TimeRange x : guestSchedule){
+            if (x.overlaps(possible)){
+              modified = true;
+              x = findIntersectionBetweenRanges(x, possible, duration);
+            }
+          }
+          if (!modified){
+            available.add(possible);
+          }
+        } else {
+          st = t.end();
+        }
+      }
+      // Checks the end of the day for availability
+      if (t != null && t.end() < LATEST_TIME && LATEST_TIME - t.end() >= duration){
+        available.add(TimeRange.fromStartEnd(t.end(), LATEST_TIME, false));
+      }
+    }
+    return available;
+  }
+
+
+  /* Finds the shared open times between multiple lists
+    @param allTimes Takes a list of each guest's schedule
+    @param duration The duration of the meeting
+    @return Returns the times where all guests are available. 
+  */
+  public List<TimeRange> findIntersectionBetweenLists(ArrayList<ArrayList<TimeRange>> allTimes, int duration){
+    List<TimeRange> intersection = new ArrayList<>();
+
+    // If there is only 1 guest, return when they are available.
+    if (allTimes.size() == 1){
+      return allTimes.get(0);
+    }
+
+    // Check each guest with the rest of the guestlist's schedule
```

Comment:
Here, you are checking the 2 guests' schedule and find available time but it doesn't cross check with other guests.
So if there are more guests it might not work.  e.g. (A,B) have free time X, (B,C) have free time Y, but X and Y might not work for all of (A,B,C).  

Good to submit now since the tests are passing. Improve if you have time.